Neos.Flow/Classes/Persistence/Doctrine/AllowedObjectsListener.php

```diff
@@ -0,0 +1,118 @@
+<?php
+namespace Neos\Flow\Persistence\Doctrine;
+
+/*
+ * This file is part of the Neos.Flow package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Doctrine\DBAL\Exception\ConnectionException;
+use Doctrine\ORM\Event\OnFlushEventArgs;
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Log\ThrowableStorageInterface;
+use Neos\Flow\Log\Utility\LogEnvironment;
+use Neos\Flow\Persistence\AllowedObjectsContainer;
+use Neos\Flow\Persistence\Exception as PersistenceException;
+use Neos\Flow\Persistence\PersistenceManagerInterface;
+use Psr\Log\LoggerInterface;
+
+/**
+ * An onFlush listener for Flow's Doctrine PersistenceManager, that validates to be persisted entities
+ * against the list of allowed objects.
+ *
+ * This listener is outsourced from the PersistenceManager to avoid recursive dependencies when building
+ * the EntityManager.
+ *
+ * @Flow\Scope("singleton")
+ * @api
+ */
+class AllowedObjectsListener
+{
+    /**
+     * @Flow\Inject
+     * @var LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * @Flow\Inject
+     * @var ThrowableStorageInterface
+     */
+    protected $throwableStorage;
+
+    /**
+     * @Flow\Inject
+     * @var AllowedObjectsContainer
+     */
+    protected $allowedObjects;
+
+    /**
+     * @Flow\Inject(lazy=true)
+     * @var PersistenceManagerInterface
```

Comment:
Yes, this is odd and somehow goes against the "recursive dependency" explanation in the class header, but we need it for `getIdentifierByObject()` in the exception message and due to the lazy (made explicit) injection works.

Reply:
Would it be better to get the PersistenceManager from the objectManager only in case of that exception rendering?

Reply:
What do you think would be the difference? As far as I can see, it would be exactly the same - but with a not so obvious alternative dependency. (ObjectManager instead of PersistenceManager)

Reply:
The OM could be used to fetch the dependency only when needed, not immediately when building the object instance.

Reply:
That's what I meant

Reply:
Yeah, would basically be a manually coded lazy dependency. Would be fine for me too, though I always cringe a bit and ask "why was this necessary instead of injection?" when using the objectmanager manually.

Reply:
After thinking a bit more about it, I would still keep the annotation. The reason is this: using the static object manager to get the instance would create a very hard coded dependency and harder to test code. Injecting the ObjectManager would couple this class to our specific DI implementation. With the annotation all this is left as a transparent implementation of our proxy compilation - which I'd like to replace with ocramius/proxymanager at some point.